src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/RSACryptoServiceProvider.cs

```diff
@@ -64,29 +65,99 @@ private RSACryptoServiceProvider(int keySize, CspParameters parameters, bool use
             // in the constructor so we can report any errors now.
             if (!_randomKeyContainer)
             {
-                GetKeyPair();
+                // Force-read the SafeKeyHandle property, which will summon it into existence.
+                SafeKeyHandle localHandle = SafeKeyHandle;
+                Debug.Assert(localHandle != null);
             }
         }
 
-        /// <summary>
-        /// Retreives the key pair
-        /// </summary>
-        private void GetKeyPair()
+        private SafeProvHandle SafeProvHandle
         {
-            if (_safeKeyHandle == null)
+            get
+            {
+                if (_safeProvHandle == null)
+                {
+                    lock (_parameters)
+                    {
+                        if (_safeProvHandle == null)
+                        {
+                            SafeProvHandle hProv = CapiHelper.CreateProvHandle(_parameters, _randomKeyContainer);
+
+                            Debug.Assert(hProv != null);
+                            Debug.Assert(!hProv.IsInvalid);
+                            Debug.Assert(!hProv.IsClosed);
+
+                            _safeProvHandle = hProv;
+                        }
+                    }
+
+                    return _safeProvHandle;
+                }
+
+                return _safeProvHandle;
+            }
+            set
             {
-                lock (this)
```

Comment:
Do we need to worry about compat issues where user code is locking on the same object?